### PR TITLE
Test Tub Detection (Assuming circular shape)

### PR DESCRIPTION
**Pull request by DavidLCPyML**
# Description

- How this works
This will take in an image feed, read and preprocess a single frame, and generate predictions of possible circular test tube openings via a Hough transform.
- Limitations
Working Hough Transform has notoriously low accuracy, can't detect plastic tube entrances unless there is a lot of contrast around the border
I suspect overfitting may be at play here since it's extremely conservative in its predictions.
- What doesn't work
Attempted RANSAC for better circle detection - didn't work as well as Hough Transform
Attempted to combine the two, but the resultant model did not yield a better result than just using a Hough Transform.
- Possible Routes
Utilizing a Neural Net-based architecture to identify test tubes via bounding boxes. Maybe YOLOv5 or some other conventional model.
# Testing
- Pictures
image

- Screenshots
WIP
- Test cases
WIP

### CODE REVIEW

File: pcr/main.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 from xarm.wrapper import XArmAPI
 from gizzmos import CustomGripper, PressureSensor
 
+print('test commit to show something')
```
Comment (sahiljain11) [inline pcr/main.py]: Do we really need this?

**Review by sahiljain11 — CHANGES_REQUESTED** [review]
This is just a test. Please add the rest of your code!

File: pcr/gizzmos.py [hunk 2]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
         return None, None
 
     def check_connection(self) -> bool:
-        return True
\ No newline at end of file
+        return True
+    
+
+import cv2 as cv
```
Comment (sahiljain11) [inline pcr/gizzmos.py:52]: Add these imports to the top of the file since that's where all of them are. Just good python coding procedure

File: pcr/gizzmos.py [hunk 3]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
...
+import numpy as np
+class TestTubeCV:
+    def __init__(self):
+        self.videoCapture = cv.VideoCapture(0, cv.CAP_DSHOW)
+        self.prevCircle = None
+        self.dist = lambda x1, y1, x2, y2: (x1 - x2) ** 2 + (y1 - y2) ** 2
+        self.debug = False
+
+    def predict(self, frame):
```
Comment (sahiljain11) [inline pcr/gizzmos.py:61]: Using python typing, you can say the following:
```
def predict(self, frame) -> List[Tuple[[int, int]]]:
```
This increases code readability so a person looking at this can more easily tell what this function takes in for input and output. You may need to import Tuple and List from typing

File: pcr/gizzmos.py [hunk 4]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
...
+            
+                self.prevCircle = chosen
+                pred.append((chosen[0], chosen[1]))
+        
+        
+        cv.imshow("circles", frame)
+        cv.imshow("grayscale", gray)
+        return pred
+
+    def getCircles(self):
```
Comment (sahiljain11) [inline pcr/gizzmos.py:104]: Same thing but you can do `def getCircles(self) -> None`

File: pcr/main.py [hunk 5]
```diff
@@ -1,15 +1,18 @@
...
 import time
 import json
 import argparse
 from mypy_types import RequestDict
 import logging
-from pcr.gizzmos import AtalantaModule
 
 from xarm.wrapper import XArmAPI
-from gizzmos import CustomGripper, PressureSensor
+from pcr.gizzmos import CustomGripper, PressureSensor, AtalantaModule
```
Comment (sahiljain11) [inline pcr/main.py:15]: You should probably add TestTubeCV instead of AtalantaModule

File: pcr/gizzmos.py [hunk 6]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
...
+        """
+        Reads in a frame from a birds' eye view video feed, predicts circles, and prints the x- and y- coordinates of the center.
+        Args: None
+        Returns: None
+        """
+        while True:
```
Comment (sahiljain11) [inline pcr/gizzmos.py:110]: Can you setup the code such as the if someone calls predict, we can just open the vidoecapture briefly (like one frame) and feed it in. Currently, this code will run forever without letting any of the xArm code to run

File: pcr/gizzmos.py [hunk 7]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
...
+        
+        # docstring of HoughCircles: HoughCircles(image, method, dp, minDist[, circles[, param1[, param2[, minRadius[, maxRadius]]]]]) -> circles
+        circles = cv.HoughCircles(gray, cv.HOUGH_GRADIENT, dp=1.0, minDist=10, 
+                                param1 = 60,
+                                param2 = 75,
+                                minRadius = 1, maxRadius = 50)
+        if circles is not None:
+            circles = np.uint16(np.around(circles))
+            chosen = None
+            for i in circles[0, :]:
```
Comment (sahiljain11) [inline pcr/gizzmos.py:87]: Could you explain in comments what's happening here? I think you're comparing the Eucledian distance between chosen and i but not sure what that means or what the algorithm here is...

File: pcr/gizzmos.py [hunk 8]
```diff
@@ -46,4 +46,78 @@ def adjust_volume(self, volume: float) -> str:
...
+            if not ret:
+                break
+
+            predicted = self.predict(frame)
+            if predicted != []:
+                print(predicted)
+
+            if cv.waitKey(1) == ord('q'):
+                break
+        self.videoCapture.release()
```
Comment (sahiljain11) [inline pcr/gizzmos.py:121]: Yeah, I don't think we need this getCircles() method tbh since it'll just be a one off call not a continuous call. I'd just use that for the predict() method. Look at the above comment for more information.

**Review by sahiljain11 — CHANGES_REQUESTED** [review]
Overall, looks pretty good! I thought the `self.dist = lambda` stuff was really cool (I'm stealing that). Listed some comments for you two iterate on! :)
